InteractiveCrowdsale/truffle/contracts/InteractiveCrowdsaleLib.sol

```diff
@@ -0,0 +1,371 @@
+    self.base.hasContributed[msg.sender] += _amount-remainder;
+
+    // make sure there are enough tokens available to satisfy the bid
+    require(numTokens <= self.base.withdrawTokensMap[self.base.owner]);
+
+    // calculate the amout of ether in the owners balance and "deposit" it
+    self.base.ownerBalance = self.base.ownerBalance + (_amount - remainder);
+
+    // add tokens to the bidders purchase.  can't overflow because it will be under the cap
+    self.base.withdrawTokensMap[msg.sender] += numTokens;
+
+    //subtract tokens from owner's share
+    (err,remainder) = self.base.withdrawTokensMap[self.base.owner].minus(numTokens);
+    require(!err);
+    self.base.withdrawTokensMap[self.base.owner] = remainder;
+
+    // add the bid to the sorted valuations list
+    uint256 listSpot;
+    listSpot = self.valuationsList.getSortedSpot(_listPredict,_personalValuation,NEXT);
+    self.valuationsList.insert(listSpot,_personalValuation,PREV);
+
+    // add the valuation to the address => valuations mapping
+    self.personalValuations[msg.sender] = _personalValuation;
+
+    // add the bidders address to the array of addresses that have submitted at the same valuation
+    self.valuationAddresses[_personalValuation].push(msg.sender);
+
+    LogBidAccepted(msg.sender, _amount-remainder, _personalValuation);
+
+    if (_currtime >= self.endWithdrawlTime) {
+      autoWithdrawBids(self);    // run algorithm to remove bids with minimal personal Valuations
+    }
+  }
+
+
+  /// @dev Called when an address wants to manually withdraw their bid from the sale. puts their wei in the LeftoverWei mapping
+  /// @return true on succesful withdrawl
+  function withdrawBid(InteractiveCrowdsaleStorage storage self) public returns (bool) {
+    // The sender has to have already bid on the sale
+    require(self.personalValuations[msg.sender] > 0);
+    // cannot withdraw after compulsory withdraw period is over
+    require(now < self.endWithdrawlTime);
+
+    // Removing the entry from the linked list returns the key of the removed entry, so make sure that was succesful
+    assert(self.valuationsList.remove(self.personalValuations[msg.sender]) == self.personalValuations[msg.sender]);
+
+    // Put the sender's contributed wei into the leftoverWei mapping for later withdrawl
+    self.base.leftoverWei[msg.sender] += self.base.hasContributed[msg.sender];
+
+    // subtract the bid from the balance of the owner
+    self.base.ownerBalance -= self.base.hasContributed[msg.sender];
+
+    // return bought tokens to the owners pool and remove tokens from the bidders pool
+    self.base.withdrawTokensMap[self.base.owner] += self.base.withdrawTokensMap[msg.sender];
+    self.base.withdrawTokensMap[msg.sender] = 0;
+
+    // iterate through the array of addresses at a certain valuation and remove the bidder's address who is withdrawing their bid
+    for (uint256 i = 0; i < self.valuationAddresses[self.personalValuations[msg.sender]].length; i++ ) {
```

Comment:
How to make sure this won't become too expensive?

Reply:
That's one of the challenges to overcome. I mentioned in the Slack. We need to think about either limiting the iterations to a number and go through via batch calls from the sale owner or track the msg.gas variable and ensure the function can operate within limits somehow.

Reply:
Ok. Is it possible to replace this array with linked list?

Reply:
we could probably replace it. still depends on the gas costs of the different operations associated. I think figuring those out is going to be my next priority